File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem52_TotalCostToHireKWorkers.java
package org.phoenix.leetcode.challenges;

import java.util.PriorityQueue;

public class Problem52_TotalCostToHireKWorkers {
    public long totalCost(int[] costs, int k, int candidates) {
        PriorityQueue<Integer> firstHalf = new PriorityQueue<>();
        PriorityQueue<Integer> lastHalf = new PriorityQueue<>();

        for (int i = 0; i < candidates; i++) {
            firstHalf.add(costs[i]);
        }
        int index = Math.max(candidates, costs.length - candidates);
        for (int i = index; i < costs.length; i++) {
            lastHalf.add(costs[i]);
        }

        int nextFirst = candidates;
        int nextLast = costs.length - candidates - 1;
        long result = 0;
        for (int i = 0; i < k; i++) {
            if (lastHalf.isEmpty() || !firstHalf.isEmpty() && firstHalf.peek() <= lastHalf.peek()) {
                result += firstHalf.poll();
                if (nextFirst <= nextLast) {
                    firstHalf.add(costs[nextFirst]);
                    nextFirst++;
                }
            } else {
                result += lastHalf.poll();
                if (nextFirst <= nextLast) {
                    lastHalf.add(costs[nextLast]);
                    nextLast--;
                }
            }
        }
        return result;
    }
}

File: JavaScript/leetcode75/src/main/Problem52_TotalCostToHireKWorkers.js
const { MinPriorityQueue } = require("@datastructures-js/priority-queue");
/**
 * @param {number[]} costs
 * @param {number} k
 * @param {number} candidates
 * @return {number}
 */
var totalCost = function (costs, k, candidates) {
  let firstHalf = new MinPriorityQueue();
  let lastHalf = new MinPriorityQueue();

  for (let i = 0; i < candidates; i++) {
    firstHalf.enqueue(costs[i]);
  }
  let index = Math.max(candidates, costs.length - candidates);
  for (let i = index; i < costs.length; i++) {
    lastHalf.enqueue(costs[i]);
  }

  let nextFirst = candidates;
  let nextLast = costs.length - candidates - 1;
  let result = 0;
  for (let i = 0; i < k; i++) {
    if (
      lastHalf.isEmpty() ||
      (!firstHalf.isEmpty() && firstHalf.front() <= lastHalf.front())
    ) {
      result += firstHalf.dequeue();
      if (nextFirst <= nextLast) {
        firstHalf.enqueue(costs[nextFirst]);
        nextFirst++;
      }
    } else {
      result += lastHalf.dequeue();
      if (nextFirst <= nextLast) {
        lastHalf.enqueue(costs[nextLast]);
        nextLast--;
      }
    }
  }
  return result;
};

module.exports = totalCost;
